crypto/finite-field.ts

```diff
@@ -266,89 +266,112 @@ function createField(
+   *
+   * It helps prevent incorrect results and infinite loops that could occur
+   * when non-canonical representations are used in operations like {@link isSquare()},
+   * {@link sqrt()}, and others. By using this function, we ensure consistent behavior
+   * across all field operations regardless of the input's initial representation.
+   */
+  function canonicalize(x: bigint) {
+    return x >= 0n && x < p ? x : mod(x, p);
+  }
+
   return {
     modulus: p,
     sizeInBits,
     t: oddFactor,
     M: twoadicity,
     twoadicRoot,
     mod(x: bigint) {
-      return mod(x, p);
+      return canonicalize(x);
     },
     add(x: bigint, y: bigint) {
-      return mod(x + y, p);
+      return canonicalize(x + y);
     },
     not(x: bigint, bits: number) {
-      return mod(2n ** BigInt(bits) - (x + 1n), p);
+      return canonicalize(2n ** BigInt(bits) - (x + 1n));
     },
     negate(x: bigint) {
-      return x === 0n ? 0n : p - x;
+      return x === 0n ? 0n : canonicalize(p - x);
     },
     sub(x: bigint, y: bigint) {
-      return mod(x - y, p);
+      return canonicalize(x - y);
     },
     mul(x: bigint, y: bigint) {
-      return mod(x * y, p);
+      return canonicalize(x * y);
     },
     inverse: exportedInverse,
     div(x: bigint, y: bigint) {
       let yinv = exportedInverse(y);
       if (yinv === undefined) return;
-      return mod(x * yinv, p);
+      return canonicalize(x * yinv);
     },
     square(x: bigint) {
-      return mod(x * x, p);
+      return canonicalize(x * x);
     },
     isSquare(x: bigint) {
-      return isSquare(x, p);
+      return isSquare(canonicalize(x), p);
     },
     sqrt(x: bigint) {
-      return sqrt(x, p, oddFactor, twoadicRoot, twoadicity);
+      const result = sqrt(
+        canonicalize(x),
+        p,
+        oddFactor,
+        twoadicRoot,
+        twoadicity
+      );
+      return result !== undefined ? canonicalize(result) : undefined;
     },
     power(x: bigint, n: bigint) {
-      return power(x, n, p);
+      return canonicalize(power(x, n, p));
     },
     dot(x: bigint[], y: bigint[]) {
       let z = 0n;
       let n = x.length;
       for (let i = 0; i < n; i++) {
-        z += x[i] * y[i];
+        z += canonicalize(x[i]) * canonicalize(y[i]);
       }
-      return mod(z, p);
+      return canonicalize(z);
     },
     equal(x: bigint, y: bigint) {
-      return mod(x - y, p) === 0n;
+      return canonicalize(x) === canonicalize(y);
     },
     isEven(x: bigint) {
-      return !(x & 1n);
+      return !(canonicalize(x) & 1n);
     },
     random() {
-      return randomField(p, sizeInBytes, hiBitMask);
+      return canonicalize(randomField(p, sizeInBytes, hiBitMask));
     },
     fromNumber(x: number) {
-      return mod(BigInt(x), p);
+      return canonicalize(BigInt(x));
     },
     fromBigint(x: bigint) {
-      return mod(x, p);
+      return canonicalize(x);
     },
     rot(
```

Comment:
`rot()`, `leftShift()` and `rightShift()` are not finite field operations, and reducing modulo the field size makes no sense here -- so please remove `canonicalize()` in inputs in these functions (they shouldn't have been put into `finite-field.ts` in the first place)

Reply:
Fixed! https://github.com/o1-labs/o1js-bindings/pull/287/commits/145a9286ab198a2156967bb8bdca6a45273827fa